enabling spmm and many corrections

RiccardoRossi opened the pull request:
**Description**
Providing CsrMatrix with the capability of not be owner of the data it works with. This is crucial for efficiency if we need to pass the data to external libraries, for example of solvers.
Right now this is implemented by manually managing the memory of "new" and by providing a user interface base on a backported implementation of std::span.

This is quite low level, and the use i make of std::span is probably a bit borderline...ideas are welcome to improve it.

Please mark the PR with appropriate tags: 
- Api Breaker 

**Changelog**
the CsrMatrix has now a flag telling if it owns the memory (read, if it is in charge of destructing it). This allow a very efficient implementation of converstion to other libraries

RiccardoRossi committed:
minor changes to expose global indices for off diagonal columns

RiccardoRossi committed:
adding value_type typedef

RiccardoRossi committed:
first tentative implementation of amgcl interface

RiccardoRossi committed:
testing that global cols indices are ok

RiccardoRossi committed:
PrintData and some other auxiliary functions

RiccardoRossi committed:
adding a test for amgcl interface

RiccardoRossi committed:
using csr_utilities to convert to amgcl

RiccardoRossi committed:
adding conversion utilities

RiccardoRossi committed:
Update kratos/mpi/utilities/distributed_csr_conversion_utilities.h

avoid using relative path for amgcl

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

RiccardoRossi committed:
Update kratos/utilities/csr_conversion_utilities.h

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

RiccardoRossi committed:
changing names of functions

RiccardoRossi committed:
renaming file

RiccardoRossi committed:
changing name

RiccardoRossi committed:
moving to the correct mpisuite and limiting test to 3 processors

RiccardoRossi committed:
fixes (by Denis) to correct when one processor is empty

RiccardoRossi committed:
fixes to enable working with empty partitions

RiccardoRossi committed:
giving correct credits to Denis Demidov

RiccardoRossi committed:
merging current master

RiccardoRossi committed:
solving conflicts and
adding exporting capabilities to amgcl matrix

RiccardoRossi committed:
added a method to SetRow and changed name from
Assemble to AssembleEntry to avoid problems with templates

RiccardoRossi committed:
adding spmm and conversion test using amgcl

RiccardoRossi committed:
adding spmm utility function

RiccardoRossi committed:
free the memory when converting amgcl to csr_matrix

RiccardoRossi committed:
minor cleanup

RiccardoRossi committed:
making the ownership of data in csr_matrix optional

loumalouomega commented on external_libraries/amgcl/mpi/distributed_matrix.hpp:
```diff
@@ -4,7 +4,7 @@
 The MIT License
 
-Copyright (c) 2012-2019 Denis Demidov <dennis.demidov@gmail.com>
+Copyright (c) 2012-2020 Denis Demidov <dennis.demidov@gmail.com>
```
2021?

RiccardoRossi replied:
next time we update the amgcl lib this will appear, but we should not change an header we don't author

loumalouomega commented on kratos/containers/csr_matrix.h:
```diff
@@ -113,7 +119,24 @@ class CsrMatrix
+        // mpRowIndicesData = nullptr;
+        // mpColIndicesData = nullptr;
+        // mpValuesVectorData = nullptr;
+        KRATOS_WATCH("entered in destructor")
```
```suggestion
```

loumalouomega commented on kratos/containers/csr_matrix.h:
```diff
@@ -113,7 +119,24 @@ class CsrMatrix
+        AssignIndex1Data(nullptr,0);
+        AssignIndex2Data(nullptr,0);
+        AssignValueData(nullptr,0);
+        KRATOS_WATCH("finished destructor")
```
```suggestion
```
And clean up commented

loumalouomega commented on kratos/containers/distributed_csr_matrix.h:
```diff
@@ -571,7 +607,21 @@ class DistributedCsrMatrix
+    virtual void PrintData(std::ostream& rOStream) const {
+
+
+        std::cout << "--- Diagonal Block: ---" << std::endl;
```
Logger?

loumalouomega commented on kratos/containers/sparse_contiguous_row_graph.h:
```diff
@@ -220,26 +221,38 @@ class SparseContiguousRowGraph
-        //sum entries
-        for(IndexType i = 1; i<static_cast<IndexType>(rRowIndices.size()); ++i){
-            rRowIndices[i] += rRowIndices[i-1];
+        if(nrows == 0) //empty
```
Style is:

~~~c++
if (condition) {
}
~~~

RiccardoRossi replied:
i formatted this with "astyle --style=allman" which is the one that i used since always... in any case, let's review this first in content, if we then agree about passing a different style formatter i will not resist

loumalouomega commented on kratos/containers/sparse_graph.h:
```diff
@@ -26,6 +26,7 @@
 // Project includes
 #include "includes/define.h"
 #include "utilities/parallel_utilities.h"
+#include <span/span.hpp>
```
If external should be in external includes, otherwise replace <> with ""

RiccardoRossi replied:
no, by using < > you are telling to consider it as a system include and not to show the eventual warnings

loumalouomega replied:
In that case should be in System includes, not in Project includes

RiccardoRossi replied:
ok

loumalouomega commented on kratos/mpi/tests/sources/test_distributed_sparse_graph.cpp:
```diff
@@ -451,6 +455,22 @@ KRATOS_DISTRIBUTED_TEST_CASE_IN_SUITE(DistributedSystemVectorConstructionMPI, Kr
+    b.SetValue(1.0);
+    pAmgcl->mul(1.0,b,1.0,y);
+
+    for(unsigned int i=0; i<y.LocalSize(); ++i)
```
Style with brackets

loumalouomega commented on kratos/tests/cpp_tests/sparse_matrices/test_sparse_graph_serial.cpp:
```diff
@@ -511,6 +517,7 @@ KRATOS_TEST_CASE_IN_SUITE(SpMV, KratosCoreFastSuite)
     for(SparseTestingInternals::IndexType i=0; i!=x.size(); ++i){
         sum += x(i);
     }
+    std::cout << " **********************" << sum << std::endl;
```
```suggestion
```
Leftover?

loumalouomega commented on kratos/utilities/amgcl_csr_conversion_utilities.h:
```diff
@@ -0,0 +1,110 @@
+        rAconverted.AssignIndex1Data((TIndexType*)(rA.ptr), rA.nrows+1);
+        rAconverted.AssignIndex2Data((TIndexType*)(rA.col), rA.nnz);
+        rAconverted.AssignValueData((TDataType*)(rA.val), rA.nnz);
+//        rAconverted.index1_data() = span<TIndexType>(&rA.ptr, rA.nrows+1);
```
Clean up this

loumalouomega commented on kratos/utilities/amgcl_csr_spmm_utilities.h:
```diff
@@ -0,0 +1,61 @@
+
+        AmgclCSRConversionUtilities::ConvertToCsrMatrix<TDataType,IndexType>(*Camgcl,rOutput);
+	}
+
```
Extra newlines

RiccardoRossi replied:
?

loumalouomega reviewed (commented):
Some comments

RiccardoRossi committed:
a bit more elegant use of span

RiccardoRossi committed:
many corrections for the case of rectangular matrix, and enabling spmm by round trip to and from amgcl routines

RiccardoRossi committed:
Update kratos/tests/cpp_tests/sparse_matrices/test_sparse_graph_serial.cpp

removing a cout

Co-authored-by: Vicente Mataix Ferrándiz <vmataix@cimne.upc.edu>

RiccardoRossi committed:
formatting

RiccardoRossi committed:
minor changes

RiccardoRossi committed:
Merge branch 'use_span_in_csr_internals' of https://github.com/KratosMultiphysics/Kratos into use_span_in_csr_internals

RiccardoRossi committed:
addressing one of vicente's comments

RiccardoRossi reviewed (commented):
ok, in my opinion this is essentially ready to go in.

this ended up being a pretty big set of changes with a lot of small corrections and new capabilities

philbucher commented on external_libraries/amgcl/adapter/zero_copy.hpp:
```diff
@@ -52,7 +52,7 @@ zero_copy(size_t n, const Ptr *ptr, const Col *col, const Val *val) {
     A->nrows = n;
     A->ncols = n;
-    A->nnz   = ptr[n];
+    A->nnz   = n ? ptr[n] : 0;
```
this seems like a delicate change
you sure abt this? Is it from upstream?

ddemidov replied:
This is also from upstream, but it does not include the recent fix for the Kratos integration:
https://github.com/ddemidov/amgcl/compare/3a207307f9b..4f260881c715

philbucher commented on external_libraries/amgcl/mpi/distributed_matrix.hpp:
```diff
@@ -1240,7 +1240,7 @@ spectral_radius(const mpi::distributed_matrix<Backend> &A, int power_iters = 0)
                 for(size_t i = 0, m = C.send.count(); i < m; ++i)
                     b0_send[i] = b0[C.send.col[i]];
-                C.exchange(&b0_send[0], &b0_recv[0]);
+                C.exchange(b0_send.data(), b0_recv.data());
```
maybe better to avoid cosmetic changes in this PR
is this from upstream ?

ddemidov replied:
It is from upstream, but it still seems to be out of sync (the line number is different):
https://github.com/ddemidov/amgcl/blob/3a207307f9b567079db6e3476c4258aad106a071/amgcl/mpi/distributed_matrix.hpp#L1224

philbucher replied:
@RiccardoRossi can you then first make a PR with the upstream changes?

RiccardoRossi replied:
well, i did not plan to update the entire library in this PR since it is always a bit problematic. I would prefer to do a PR AFTER this PR is merged

RiccardoRossi replied:
anyhow, this is not cosmetic, it is needed in the case in which the array is empty

philbucher commented on kratos/tests/cpp_tests/sparse_matrices/run_my_test.py:
```diff
@@ -9,7 +9,8 @@
+Tester.RunTestCases("TestToAMGCLMatrix")
 
 #Tester.RunTestCases("TestPerformanceBenchmarkSparseGraph")
 #Tester.RunTestCases("TestPerformanceBenchmarkSparseContiguousRowGraph")
```
please remove this file, it should not be in master

philbucher commented on kratos/utilities/amgcl_csr_conversion_utilities.h:
```diff
@@ -0,0 +1,82 @@
+{
+
+/**
+Utilities to convert the distributed_csr matrix to other libraries
```
```suggestion
Utilities to convert the distributed_csr matrix to an AMGCL matrix
```

philbucher commented on kratos/utilities/amgcl_csr_spmm_utilities.h:
```diff
@@ -0,0 +1,60 @@
+        const CsrMatrix<TDataType, TIndexType>& rB
+        )
+	{
+        //bool move_to_backend=false;
```
```suggestion
```

philbucher commented on kratos/utilities/amgcl_csr_spmm_utilities.h:
```diff
@@ -0,0 +1,60 @@
+     This function returns a shared pointer to an Amgcl distributed_matrix
+     */
+    template< class TDataType=double, class TIndexType=std::size_t >
+    static CsrMatrix<TDataType, TIndexType> SparseMultiply(
```
what is the purpose of this?
I thought you implemented this also directly for the Kratos Matrix?

RiccardoRossi replied:
no, the SpMM is way more complicated than SpMV. I am using here the implementation in amgcl which is quite optimized. Note  that an important point of this PR is exactly to have a very fast conversion to and from AMGCL (although one could use different libs)

philbucher commented on kratos/tests/cpp_tests/sparse_matrices/test_sparse_graph_serial.cpp:
```diff
@@ -552,6 +558,120 @@ KRATOS_TEST_CASE_IN_SUITE(SystemVectorOperations, KratosCoreFastSuite)
+
+    std::vector<double> x(A.size1());
+    for(auto& item : x)
+       item = 1.0;
```
```suggestion
    std::vector<double> x(A.size1(), 1.0);
```

RiccardoRossi replied:
well, this should be 1, but ok

philbucher commented on kratos/tests/cpp_tests/sparse_matrices/test_sparse_graph_serial.cpp:
```diff
@@ -552,6 +558,120 @@ KRATOS_TEST_CASE_IN_SUITE(SystemVectorOperations, KratosCoreFastSuite)
+
+    std::vector<double> y(A.size1());
+    for(auto& item : y)
+       item = 0.0;
```
```suggestion
    std::vector<double> y(A.size1(), 0.0);
```

philbucher commented on kratos/tests/cpp_tests/sparse_matrices/test_sparse_graph_serial.cpp:
```diff
@@ -308,10 +311,13 @@ KRATOS_TEST_CASE_IN_SUITE(GraphConstruction, KratosCoreFastSuite)
-    SparseTestingInternals::CheckCSRGraphArrays(row_index, col_index, reference_A_map);
+    SparseTestingInternals::CheckCSRGraphArrays(row_index_span, col_index_span, reference_A_map);
+    delete [] row_index_span.begin();
+    delete [] col_index_span.begin();
```
so this is related to the spans?
I thought they don't own / allocate memory?

RiccardoRossi replied:
that's exactly the point. They do not own the memory...so i need to explicitly delete them (the pointer they have to the first entry is valid)

RiccardoRossi replied:
note that i don't love this way of using. This is the reason for which i also created a version on which one explicitly needs to pass raw pointers to the function

philbucher commented on kratos/utilities/amgcl_csr_conversion_utilities.h:
```diff
@@ -0,0 +1,82 @@
+        Aconverted.AssignIndex1Data((TIndexType*)(rA.ptr), rA.nrows+1);
+        Aconverted.AssignIndex2Data((TIndexType*)(rA.col), rA.nnz);
+        Aconverted.AssignValueData((TDataType*)(rA.val), rA.nnz);
+        return Aconverted;
```
no `std::move` here?
You use it in the MPI-equivalent

RiccardoRossi replied:
well, i get the following if i add it

~~~
In file included from /mnt/c/Users/rross/Programming/Kratos/kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h:24,
                 from /mnt/c/Users/rross/Programming/Kratos/kratos/mpi/tests/sources/test_distributed_sparse_graph.cpp:28:
/mnt/c/Users/rross/Programming/Kratos/kratos/utilities/amgcl_csr_conversion_utilities.h: In instantiation of ‘static Kratos::CsrMatrix<TDataType, TIndexType> Kratos::AmgclCSRConversionUtilities::ConvertToCsrMatrix(typename amgcl::backend::builtin<T1>::matrix&) [with TDataType = double; TIndexType = long unsigned int; typename amgcl::backend::builtin<T1>::matrix = amgcl::backend::crs<double, long int, long int>]’:
/mnt/c/Users/rross/Programming/Kratos/kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h:113:114:   required from ‘static Kratos::DistributedCsrMatrix<TDataType, TIndexType> Kratos::AmgclDistributedCSRConversionUtilities::ConvertToCsrMatrix(amgcl::mpi::distributed_matrix<amgcl::backend::builtin<double> >&) [with TDataType = double; TIndexType = long unsigned int]’
/mnt/c/Users/rross/Programming/Kratos/kratos/mpi/tests/sources/test_distributed_sparse_graph.cpp:493:109:   required from here
/mnt/c/Users/rross/Programming/Kratos/kratos/utilities/amgcl_csr_conversion_utilities.h:73:36: warning: moving a local object in a return statement prevents copy elision [-Wpessimizing-move]
   73 |         return std::move(Aconverted);
      |                                    ^
/mnt/c/Users/rross/Programming/Kratos/kratos/utilities/amgcl_csr_conversion_utilities.h:73:36: note: remove ‘std::move’ call
~~~

philbucher commented on kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h:
```diff
@@ -0,0 +1,158 @@
+
+		DistributedCsrMatrix<TDataType, TIndexType> Aconverted;
+
+		DataCommunicator& kratos_comm=ParallelEnvironment::GetDefaultDataCommunicator();
```
the default data comm should really not be used internally!
it is already scheduled to speak about this in the KTC

Looking at the function I would say you need to pass it from outside

RiccardoRossi replied:
well, i can see your point, however one would have to create it here on the basis of a raw communicator (what the amgcl uses) and i don't think this is supported right now

philbucher commented on kratos/mpi/tests/sources/test_distributed_sparse_graph.cpp:
```diff
@@ -619,5 +667,177 @@ KRATOS_DISTRIBUTED_TEST_CASE_IN_SUITE(DistributedSystemVectorOperationsMPI, Krat
+KRATOS_TEST_CASE_IN_SUITE(SmallRectangularDistributedMatrixMatrixMultiply, KratosMPICoreFastSuite)
+{
+    typedef std::size_t IndexType;
+    DataCommunicator& rComm=ParallelEnvironment::GetDefaultDataCommunicator();
```
for testing this is fine

philbucher commented on kratos/containers/sparse_graph.h:
```diff
@@ -187,59 +188,79 @@ class SparseGraph
-        {
-            rRowIndices.resize(nrows+1, false);
-        }
+        pRowIndicesData = new IndexType[nrows+1];
```
you add `new` but no `delete` anwhere
can you explain?

RiccardoRossi replied:
yes, the delete needs to be done by the class that calls this function.

philbucher commented on kratos/containers/sparse_graph.h:
```diff
@@ -187,59 +188,79 @@ class SparseGraph
-        }
+        IndexType nnz = row_indices[nrows];
+        rColDataSize = nnz;
+        pColIndicesData = new IndexType[nnz];
```
same

RiccardoRossi replied:
same as before. ownership passed to the caller

philbucher replied:
adding a comment abt this wouldn't hurt ;)

philbucher commented on kratos/containers/sparse_contiguous_row_graph.h:
```diff
@@ -209,58 +210,84 @@ class SparseContiguousRowGraph
-        }
+        IndexType nnz = row_indices[nrows];
+        rColDataSize=nnz;
+        pColIndicesData = new IndexType[nnz];
```
same

RiccardoRossi replied:
the ownership is transferred to the caller, who is in charge of deleting those arrays...

philbucher commented on kratos/containers/distributed_csr_matrix.h:
```diff
@@ -69,17 +69,20 @@ class DistributedCsrMatrix
     ///@name Life Cycle
     ///@{
+    DistributedCsrMatrix() //default constructor. To be used for low level operations
+        : mrComm(ParallelEnvironment::GetDefaultDataCommunicator())
```
I would not allow this
this can lead to nasty deadlocks
IMO the user must know which data-comm to use

philbucher commented on kratos/containers/distributed_csr_matrix.h:
```diff
@@ -247,60 +328,114 @@ class DistributedCsrMatrix
 
+    std::map<IndexType, IndexType>& GetOffDiagonalLocalIds()
+    {
+        return mOffDiagonalLocalIds; //usage: mOffDiagonalLocalIds[global_id] contains the local_id associated to that global_id (for a off diagonal block entry)
```
those comment should go to the doc of the function

RiccardoRossi replied:
i disagree ... it is good that it is also here

philbucher commented on kratos/containers/csr_matrix.h:
```diff
@@ -567,18 +713,60 @@ class CsrMatrix
+        }
+
+        IndexType rows_size;
+        //rSerializer.load("mpRowIndicesData",mpRowIndicesData);
```
?

philbucher commented on kratos/containers/csr_matrix.h:
```diff
@@ -567,18 +713,60 @@ class CsrMatrix
+            rSerializer.load("i",mRowIndices[i]);
+
+        IndexType cols_size;
+        //rSerializer.load("mpColIndicesData",mpColIndicesData);
```
?

philbucher commented on kratos/containers/csr_matrix.h:
```diff
@@ -567,18 +713,60 @@ class CsrMatrix
+            rSerializer.load("i",mColIndices[i]);
+
+        IndexType vals_size;
+        //rSerializer.load("mpValuesVectorData",mpValuesVectorData);
```
?

philbucher commented on kratos/containers/csr_matrix.h:
```diff
@@ -475,7 +601,23 @@ class CsrMatrix
     /// Print object's data.
-    virtual void PrintData(std::ostream& rOStream) const {}
+    virtual void PrintData(std::ostream& rOStream) const {
+        std::cout << "size1 : " << size1() <<std::endl;
```
logger

philbucher commented on kratos/containers/csr_matrix.h:
```diff
@@ -111,21 +117,77 @@ class CsrMatrix
+        // mpValuesVectorData = nullptr
+        AssignIndex1Data(nullptr,0);
+        AssignIndex2Data(nullptr,0);
+        AssignValueData(nullptr,0);        
```
This is confusing, who deleted the data?

RiccardoRossi replied:
the Assign functions replace the data and do the deletion internally. Actually this was one of the hard parts of the PR

philbucher reviewed (commented):
you are changing the amgcl code
are those changes from upstream?
If so then I would suggest to update amgcl entirely in an upfront PR

also I didn't see test for Serialization

furthermore I pointed out some (IMO) wrong usages of the DataCommunicator

RiccardoRossi committed:
Update csr_matrix.h

removing a watch and some unneeded lines

RiccardoRossi committed:
Update kratos/utilities/amgcl_csr_conversion_utilities.h

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

RiccardoRossi committed:
Apply suggestions from code review

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

RiccardoRossi committed:
Update kratos/tests/cpp_tests/sparse_matrices/test_sparse_graph_serial.cpp

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

RiccardoRossi committed:
addressing PR comments

RiccardoRossi reviewed (commented):
some answers to the PR comments

RiccardoRossi committed:
Merge branch 'master' into use_span_in_csr_internals

RiccardoRossi committed:
adding missing include

RiccardoRossi committed:
Merge branch 'master' into use_span_in_csr_internals

RiccardoRossi committed:
moving a test to the correct suite

RiccardoRossi committed:
Merge branch 'master' into use_span_in_csr_internals

sunethwarna committed:
Merge remote-tracking branch 'origin/master' into use_span_in_csr_internals

RiccardoRossi committed:
Merge branch 'master' into use_span_in_csr_internals

RiccardoRossi committed:
cleanup in the use of span

sunethwarna committed:
changing the clear method

roigcarlo committed:
Enable full test output

RiccardoRossi committed:
attempting to fix problem in win

RiccardoRossi committed:
minimal correction

RiccardoRossi committed:
recorrecting a wrong correction

roigcarlo committed:
Undo changes

RiccardoRossi committed:
minor improvements

roigcarlo committed:
Inheriting ownership from move constructor

This should fix the windows problems. Still investigating why the move behavior is different between platforms

roigcarlo committed:
Previnting use of elated constructor in move returns

RiccardoRossi committed:
using pointers to prevent problems with std move

RiccardoRossi committed:
Merge branch 'use_span_in_csr_internals' of https://github.com/KratosMultiphysics/Kratos into use_span_in_csr_internals

RiccardoRossi committed:
small cleanup

RiccardoRossi commented on kratos/containers/distributed_csr_matrix.h:
```diff
@@ -913,8 +922,8 @@ class DistributedCsrMatrix
 
-    CsrMatrix<TDataType,IndexType> mDiagonalBlock;
-    CsrMatrix<TDataType,IndexType> mOffDiagonalBlock;
+    typename CsrMatrix<TDataType,IndexType>::Pointer mpDiagonalBlock;
```
@roigcarlo  possibly it would be best to use a unique_ptr here

RiccardoRossi commented on kratos/containers/csr_matrix.h:
```diff
@@ -151,8 +151,8 @@ class CsrMatrix
     //move assignement operator
     CsrMatrix& operator=(CsrMatrix&& rOtherMatrix)
     {
+        mIsOwnerOfData=rOtherMatrix.mIsOwnerOfData;
```
@roigcarlo this was also missing (please comment)

roigcarlo replied:
Yep, :+1:

RiccardoRossi committed:
corrections. should now pass tests

RiccardoRossi committed:
trying to make centos old gcc happy
(code should be already valid)

RiccardoRossi committed:
next attempt at making centos happy

RiccardoRossi committed:
using TIndexType everywhere to attempt
making old gcc happy

loumalouomega committed:
Using format (I)

loumalouomega committed:
Using format (II)

loumalouomega committed:
Using format (III)

loumalouomega committed:
It is external

loumalouomega committed:
Using format (IV)

loumalouomega committed:
Proper format

roigcarlo reviewed (commented):
Looks good to me after review, and problems seem solved so :+1:

RiccardoRossi committed:
Update kratos/utilities/amgcl_csr_spmm_utilities.h

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

RiccardoRossi committed:
removign unneeded file

RiccardoRossi committed:
Merge branch 'use_span_in_csr_internals' of https://github.com/KratosMultiphysics/Kratos into use_span_in_csr_internals

RiccardoRossi committed:
Merge branch 'master' into use_span_in_csr_internals

loumalouomega committed:
Correct

loumalouomega committed:
Minor

RiccardoRossi committed:
Merge branch 'use_span_in_csr_internals' of https://github.com/KratosMultiphysics/Kratos into use_span_in_csr_internals

rubenzorrilla commented on kratos/containers/sparse_contiguous_row_graph.h:
```diff
@@ -209,58 +209,101 @@ class SparseContiguousRowGraph
+    IndexType ExportCSRArrays(
+        span<IndexType>& rRowIndices,
+        span<IndexType>& rColIndices
+    ) const = delete;
```
Just for me to know, why deleting instead of not declaring it?

RiccardoRossi replied:
span does not transfer ownership of the memory, so this would give a memory error. To prevent it i am explicitly deleting it

rubenzorrilla commented on kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h:
```diff
@@ -0,0 +1,167 @@
+        // else{ //if rA is the owner, transfer ownership to the csr_matrix
+        //     rA.local()->own_data = false;
+        //     pAconverted->GetOffDiagonalBlock().SetIsOwnerOfData(true);
+        // }		
```
Is this intentionally kept?

RiccardoRossi replied:
i can remove much of it, however i would prefer to merge and to have another small PR just taking it out

philbucher commented on kratos/tests/cpp_tests/sparse_matrices/test_sparse_graph_serial.cpp:
```diff
@@ -201,8 +204,8 @@ bool CheckGraph(
-    const vector<SparseTestingInternals::IndexType>& rRowIndices,
-    const vector<SparseTestingInternals::IndexType>& rColIndices,
+    const span<SparseTestingInternals::IndexType>& rRowIndices,
+    const span<SparseTestingInternals::IndexType>& rColIndices,
```
how about wrapping this in `Kratos::span` like e.g. `shared_ptr`
=> this will make the transition to C++20 `std::span` much easier

RiccardoRossi replied:
it is actually done that way, however since "span" is added to the Kratos namespace it is not needed to add the Kratos:: in front of it

RiccardoRossi replied:
anyhow, i'll add the Kratos:: everywhere (it is quickly done)

philbucher commented on kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h:
```diff
@@ -0,0 +1,167 @@
+	template< class TDataType, class TIndexType >
+	static typename DistributedCsrMatrix<TDataType, TIndexType>::UniquePointer ConvertToCsrMatrix(
+			amgcl::mpi::distributed_matrix<amgcl::backend::builtin<double>>& rA, //cannot be made const since i need to modify some data in-place,
+			DataCommunicator& kratos_comm=ParallelEnvironment::GetDefaultDataCommunicator()
```
```suggestion
			DataCommunicator& kratos_comm
```

this is dangerous, can you please remove it?

RiccardoRossi replied:
well, i would say that it is a fair policy to pass the default, as long as a user can change it as he wishes. Having said this it is little to no work to change it so i have done it

philbucher reviewed (commented):
looks good to me

I made two important comments, please take a look (related to `span` and usage of `DefaultDataComm`)

RiccardoRossi committed:
making explicit that we are using Kratos::span

RiccardoRossi committed:
minor cleanups

RiccardoRossi committed:
avoiding passing default_comm

RiccardoRossi commented on kratos/containers/sparse_contiguous_row_graph.h:
```diff
@@ -209,58 +209,101 @@ class SparseContiguousRowGraph
+    ) const = delete;
+
+    IndexType ExportCSRArrays(
+        IndexType*& pRowIndicesData,
```
just to tell that i would have loved to manage the memory by a shared_ptr<TDataType[]> but this is only supported by c++17 or higher.
also i am not completely sure if this would have no performance implications.

actually an even better option would have been to use 
unique_ptr<TDataType[]> but this has difficulties since we may have matrices who do not own the pointer
